krun/vm_defs.py

```diff
@@ -480,12 +579,105 @@ def __init__(self, vm_path, env=None):
         lib_dir = os.path.dirname(vm_path)
         self.add_env_change(EnvChangeAppend("LD_LIBRARY_PATH", lib_dir))
 
+    def parse_instrumentation_stderr_file(self, file_handle):
+        """PyPy instrumentation data collected from the PYPYLOG.
+
+        We collect:
+          * The time spent in compilation-related tasks.
+          * The time spent in GC-related tasks.
+
+        Note that the times are not in wall-clock time. Consider the units
+        arbitrary.
+
+        Events are not necessarily sequential. They can be nested. E.g. GC can
+        happen inside tracing. The upshot is: we have to be a bit clever about
+        how we separate GC time from compilation time.
+
+        We treat the event stream like a tree. We walk this tree, keeping track
+        of the nesting of events. When computing the time for each event, we
+        exclude the time spent in nested events. The nested children have their
+        times computed separately.
+
+        The INSTRUMENTATION mapping decides which kind of events contribute
+        to which time counter (gc or compilation).
+        """
+
+        # This stores the counters for all in-process iterations
+        data = {
+            "gc_time": [],
+            "compilation_time": [],
+        }
+
+        # This stores the counters for the current in-process iteration
+        data_this_iter = {
+            "gc_time": 0,
+            "compilation_time": 0
+        }
+
+        iter_num = 0
+        current_event = None
+
```

Comment:
Do you mean butt the variable up against the loop header? I would prefer to keep this white-space for readability.